### DISCUSSION TITLE
Which metadata extension is run first?

### DISCUSSION THREAD

**Discussion by andrewpbray** [quarto-dev/quarto-cli#10598] [Q&A] [answered]
### Description

If I add two different [metadata extensions](https://quarto.org/docs/extensions/metadata.html) to my Quarto project, each with a pre-render script, which one runs first?

This brings to my mind something that's been slowly maturing for me over the months/years. Lua Filters modify the way Quarto/pandoc compiles single docs. They're effectively Document Extensions. A document can call many document extensions and specify the order in which they're run.

Pre-and post-render scripts modify the way Quarto compiles multidoc projects. They're effectively Project Extensions. A project can call many such project extensions, but ... I don't know of a way to specify the order they're run (or make apparent in `_quarto.yml` that they're even being run).

The analog to `filters` appearing in document yaml seems like it'd be a `_quarto.yml` that read:

```yml
project:
  type: website
  extensions:
    - my-first-extension
    - my-second-extension
```

**Comment by cscheid** [comment, marked as answer]
I hear you.

Context: We designed the filter declarations purposefully expecting that they would interfere with one another and so the user needs to declare their order. We generally don't expect extensions to interfere with one another, and so we didn't want to burden users with the same verboseness that comes in filter declarations.

Technical answer: we specify that behavior (that is, you can't really currently control it, because we don't promise it'll be a defined order).

Honestly: I don't know, we'd have to look. I don't think I _want_ users to be able to control it, because I'd like to encourage extensions to be written in a way that they _are_ independent of each other.

**Reply by andrewpbray** [reply]
The independence/exchangability makes sense.

The other use-case I'm running up against isn't order of extensions, but inclusion/exclusion. I have a case where I'd like one project profile to use a project extension and second project profile to not use it.

There are workarounds: hard-coding paths to scripts in the profile` _quarto.yml` and for some, including or excluding `metadata-files`, but both break the wall of the fully encapsulated extension.

**Reply by cscheid** [reply]
> I have a case where I'd like one project profile to use a project extension and second project profile to not use it.

> but both break the wall of the fully encapsulated extension

I think I still don't quite understand the use case. If the project profiles are expected to be different, does adding the extension lists under each project profile not work?

**Reply by andrewpbray** [reply]
> If the project profiles are expected to be different, does adding the extension lists under each project profile not work?

Hmm - I'm not aware of how to add an extension list to a project profile. Do you mean specifying the paths in to the scripts in `_extensions` under `project.pre-render`?

**Reply by cscheid** [reply]
... nevermind - I read your proposed syntax and it infiltrated my mind! I guess this is an admission that we should have that :)

**Reply by fuhrmanator** [reply]
I have a similar question -- I'm using the fontawesome extension, and I wanted to write some lua filter to add some icons into certain styled blocks.  I was hoping to just insert inline markdown `{{< fa thumbs-up >}}`

But in debugging my filter it seems that the `{{< fa thumbs-up >}}` (or similar) has already been converted to Span with __quarto_custom_type, etc. before my filter runs:

```
Span ("",[],[("__quarto_custom","true"),("__quarto_custom_type","Shortcode"),("__quarto_custom_context","Inline"),("__quarto_custom_id","1")]) []]
```

I tried putting a full path to the fontawesome.lua (filter?) from the extension at the bottom of my `filters:` list, but it changes nothing (and doesn't error out).

If I understand it, we can't control this order either? Anyway, maybe it's another use case.

**Reply by cscheid** [reply]
@fuhrmanator I think you're trying to manipulate the contents of a shortcode expression at runtime, or, equivalently, to create a new shortcode expression in Lua. 

It takes a bit of trickery, but your filter can target the AST node for the Shortcode directly. We envisioned this to exist for people to replace the behavior of certain shortcodes; you want to, in addition, create a _new_ shortcode from Lua, which is a bit harder. But this is a minimal example:


````qmd
---
format: html
key1: value1
key2: value2
filters:
  - change_meta_shortcode.lua
---

{{< meta key1 >}}
````

The filter:

```lua
-- change_meta_shortcode.lua
function Shortcode(shortcode)
  local node = quarto._quarto.ast.create_custom_node_scaffold("Shortcode", "Inline")
  return quarto.Shortcode({
    __quarto_custom_node = node,
    name = "meta",
    params = {
      {
        type = "param",
        value = "key2"
      }
    }
  })
end
```

**Reply by cscheid** [reply]
For a bit more context: the `Span` you're seeing is the internal representation of a custom AST node for shortcodes in Quarto's AST. (We need it to be like that to avoid Pandoc interpreting shortcode parameter strings as Markdown: `{{< eg this--should--not--have--en--dashes >}}` otherwise behaves badly). Quarto actually "transforms" those raw shortcodes before Pandoc sees them. Then, our internal Lua representation converts that transformed input into a custom AST node, and from there, filters can target it.

This is not the first time someone asked for the ability to create shortcode invocations from inside a Lua filter. We should allow that: tracking the enhancement request [here](https://github.com/quarto-dev/quarto-cli/issues/11501).

**Reply by fuhrmanator** [reply]
> This is not the first time someone asked for the ability to create shortcode invocations from inside a Lua filter.

Thanks for the info and the enhancement request. 

I finally used macros (\newcommand) in LaTeX, since it's possible to put them in the quarto source. Of course it only works in PDF format. It's the old way I used to do this before shortcodes and the fontawesome extension :)
